File: apps/backend/src/services/task_service.py
import json
import sqlite3
import uuid
from datetime import datetime

from src.db.db_manager import db_manager


class TaskService:
    def _get_conn(self):
        return sqlite3.connect(db_manager.get_db_path())

    def create_task(
        self,
        title,
        platforms,
        file_list,
        account_list,
        schedule_data,
        publish_data=None,
        priority=1,
    ):
        task_id = f"task_{int(datetime.now().timestamp())}_{str(uuid.uuid4())[:8]}"
        conn = self._get_conn()
        cursor = conn.cursor()
        try:
            cursor.execute(
                """
                INSERT INTO tasks (id, title, status, progress, priority, platforms, file_list, account_list, schedule_data, publish_data)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
                (
                    task_id,
                    title,
                    "waiting",  # Initial status
                    0,
                    priority,
                    json.dumps(platforms),
                    json.dumps(file_list),
                    json.dumps(account_list),
                    json.dumps(schedule_data),
                    json.dumps(publish_data) if publish_data else None,
                ),
            )
            conn.commit()
            return task_id
        except Exception as e:
            print(f"创建任务失败: {e}")
            conn.rollback()
            return None
        finally:
            conn.close()

    def update_task_status(self, task_id, status, progress=None, error_msg=None):
        conn = self._get_conn()
        cursor = conn.cursor()
        try:
            updates = ["status = ?", "updated_at = CURRENT_TIMESTAMP"]
            params = [status]

            if progress is not None:
                updates.append("progress = ?")
                params.append(progress)

            if error_msg is not None:
                updates.append("error_msg = ?")
                params.append(error_msg)

            params.append(task_id)

            sql = f"UPDATE tasks SET {', '.join(updates)} WHERE id = ?"
            cursor.execute(sql, params)
            conn.commit()
        except Exception as e:
            print(f"更新任务状态失败: {e}")
        finally:
            conn.close()

    def get_all_tasks(self):
        conn = self._get_conn()
        conn.row_factory = sqlite3.Row
        cursor = conn.cursor()
        try:
            cursor.execute("SELECT * FROM tasks ORDER BY created_at DESC")
            rows = cursor.fetchall()
            tasks = []
            for row in rows:
                task = dict(row)
                # Parse JSON fields
                try:
                    task["platforms"] = json.loads(task["platforms"]) if task["platforms"] else []
                except:
                    pass
                try:
                    task["file_list"] = json.loads(task["file_list"]) if task["file_list"] else []
                except:
                    pass
                try:
                    task["account_list"] = (
                        json.loads(task["account_list"]) if task["account_list"] else []
                    )
                except:
                    pass
                # Parse schedule_data
                try:
                    task["schedule_data"] = (
                        json.loads(task["schedule_data"]) if task["schedule_data"] else {}
                    )
                except:
                    task["schedule_data"] = {}
                try:
                    task["publish_data"] = (
                        json.loads(task["publish_data"]) if task["publish_data"] else {}
                    )
                except:
                    task["publish_data"] = {}
                # Clean up schedule_data if needed
                tasks.append(task)
            return tasks
        finally:
            conn.close()

    def delete_task(self, task_id):
        conn = self._get_conn()
        cursor = conn.cursor()
        try:
            cursor.execute("DELETE FROM tasks WHERE id = ?", (task_id,))
            conn.commit()
            return True
        except Exception as e:
            print(f"删除任务失败: {e}")
            return False
        finally:
            conn.close()


task_service = TaskService()

File: apps/backend-node/src/services/task-service.ts
/**
 * Task service for omni-post backend (Node.js).
 * Mirrors: apps/backend/src/services/task_service.py
 */

import { v4 as uuidv4 } from 'uuid';
import { SpanStatusCode } from '@opentelemetry/api';
import { logger } from '../core/logger.js';
import { getTracer } from '../core/telemetry.js';
import { dbManager } from '../db/database.js';

export interface Task {
    id: string;
    title: string | null;
    status: string;
    progress: number;
    priority: number;
    platforms: number[];
    file_list: string[];
    account_list: string[];
    schedule_data: any;
    error_msg: string | null;
    publish_data: any;
    created_at: string;
    updated_at: string;
}

/**
 * Task service class for managing task lifecycle.
 */
class TaskService {
    /**
     * Create a new publishing task.
     */
    createTask(publishData: any): string {
        const tracer = getTracer();
        return tracer.startActiveSpan('task.create', (span) => {
            try {
                const db = dbManager.getDb();
                const taskId = `task_${Date.now()}_${uuidv4().slice(0, 8)}`;
                span.setAttribute('task.id', taskId);

                const stmt = db.prepare(`
      INSERT INTO tasks (id, title, status, progress, priority, platforms, file_list, account_list, schedule_data, publish_data)
      VALUES (?, ?, 'waiting', 0, 1, ?, ?, ?, ?, ?)
    `);

                // Build schedule_data from top-level fields (matching Python's route-layer assembly)
                const scheduleData = {
                    enableTimer: publishData.enableTimer ?? null,
                    videosPerDay: publishData.videosPerDay ?? null,
                    dailyTimes: publishData.dailyTimes ?? null,
                    startDays: publishData.startDays ?? null,
                };

                stmt.run(
                    taskId,
                    publishData.title || null,
                    JSON.stringify(publishData.platforms || [publishData.type]),
                    JSON.stringify(publishData.fileList || []),
                    JSON.stringify(publishData.accountList || []),
                    JSON.stringify(scheduleData),
                    JSON.stringify(publishData)
                );

                return taskId;
            } catch (error: unknown) {
                const exception = error instanceof Error ? error : new Error(String(error));
                span.recordException(exception);
                span.setStatus({ code: SpanStatusCode.ERROR, message: exception.message });
                throw exception;
            } finally {
                span.end();
            }
        }); // end startActiveSpan
    }

    /**
     * Update task status and progress.
     * Includes protection logic: 'uploading' updates won't overwrite 'completed' or 'failed' status.
     */
    updateTaskStatus(
        taskId: string,
        status: string,
        progress?: number,
        errorMsg?: string | null,
        force = false
    ): void {
        const tracer = getTracer();
        tracer.startActiveSpan('task.updateStatus', (span) => {
            try {
                span.setAttribute('task.id', taskId);
                span.setAttribute('task.status', status);
                if (progress !== undefined) span.setAttribute('task.progress', progress);

                const db = dbManager.getDb();
                const forceNum = force ? 1 : 0;

                // Use a single UPDATE with conditional WHERE to enforce guards:
                // 1. Force bypass (? = 1)
                // 2. Restart guard: ((status NOT IN ('completed', 'failed') OR (? = 0 AND ? IN ('waiting', 'uploading'))))
                //    Allows moving out of terminal states ONLY for explicit restarts (progress 0 + non-terminal status)
                // 3. Progress guard: (status != ? OR ? >= progress)
                //    Prevents progress regression when status remains the same

                let stmt;
                let params: (string | number | null)[];

                if (progress !== undefined && errorMsg !== undefined) {
                    stmt = db.prepare(`
                UPDATE tasks
                SET status = ?, progress = ?, error_msg = ?, updated_at = CURRENT_TIMESTAMP
                WHERE id = ? AND (
                    ? = 1 OR (
                        (status NOT IN ('completed', 'failed') OR (? = 0 AND ? IN ('waiting', 'uploading'))) AND
                        (status != ? OR ? >= progress)
                    )
                )
            `);
                    params = [status, progress, errorMsg, taskId, forceNum, progress, status, status, progress];
                } else if (progress !== undefined) {
                    stmt = db.prepare(`
                UPDATE tasks
                SET status = ?, progress = ?, updated_at = CURRENT_TIMESTAMP
                WHERE id = ? AND (
                    ? = 1 OR (
                        (status NOT IN ('completed', 'failed') OR (? = 0 AND ? IN ('waiting', 'uploading'))) AND
                        (status != ? OR ? >= progress)
                    )
                )
            `);
                    params = [status, progress, taskId, forceNum, progress, status, status, progress];
                } else if (errorMsg !== undefined) {
                    stmt = db.prepare(`
                UPDATE tasks
                SET status = ?, error_msg = ?, updated_at = CURRENT_TIMESTAMP
                WHERE id = ? AND (
                    ? = 1 OR (status NOT IN ('completed', 'failed'))
                )
            `);
                    params = [status, errorMsg, taskId, forceNum];
                } else {
                    stmt = db.prepare(`
                UPDATE tasks
                SET status = ?, updated_at = CURRENT_TIMESTAMP
                WHERE id = ? AND (
                    ? = 1 OR (status NOT IN ('completed', 'failed'))
                )
            `);
                    params = [status, taskId, forceNum];
                }

                const info = stmt.run(...params);
                if (info.changes === 0) {
                    // Either ID doesn't exist OR it was blocked by guards
                    // We can't easily distinguish without another SELECT, but we can log that it was skipped
                    const progressInfo = progress !== undefined ? ` (${progress}%)` : '';
                    logger.debug(`[TaskService] Task update skipped (guarded or non-existent): ${taskId} -> ${status}${progressInfo}`);
                }
            } catch (error: unknown) {
                const exception = error instanceof Error ? error : new Error(String(error));
                span.recordException(exception);
                span.setStatus({ code: SpanStatusCode.ERROR, message: exception.message });
                throw exception;
            } finally {
                span.end();
            }
        }); // end startActiveSpan
    }

    /**
     * Get all tasks, sorted by creation time (newest first).
     */
    getAllTasks(): Task[] {
        const db = dbManager.getDb();
        const rows = db.prepare('SELECT * FROM tasks ORDER BY created_at DESC').all() as any[];

        return rows.map(row => ({
            ...row,
            platforms: this._parseJson(row.platforms, []),
            file_list: this._parseJson(row.file_list, []),
            account_list: this._parseJson(row.account_list, []),
            schedule_data: this._parseJson(row.schedule_data, null),
            publish_data: this._parseJson(row.publish_data, null),
        }));
    }

    /**
     * Get a task by ID.
     */
    getTask(taskId: string): Task | null {
        const db = dbManager.getDb();
        const row = db.prepare('SELECT * FROM tasks WHERE id = ?').get(taskId) as any;

        if (!row) return null;

        return {
            ...row,
            platforms: this._parseJson(row.platforms, []),
            file_list: this._parseJson(row.file_list, []),
            account_list: this._parseJson(row.account_list, []),
            schedule_data: this._parseJson(row.schedule_data, null),
            publish_data: this._parseJson(row.publish_data, null),
        };
    }

    /**
     * Delete a task by ID.
     */
    deleteTask(taskId: string): boolean {
        const db = dbManager.getDb();
        const result = db.prepare('DELETE FROM tasks WHERE id = ?').run(taskId);
        return result.changes > 0;
    }

    /**
     * Safe JSON parsing with fallback.
     */
    private _parseJson(value: string | null, fallback: any): any {
        if (!value) return fallback;
        try {
            return JSON.parse(value);
        } catch {
            return fallback;
        }
    }
}

// Global singleton instance
export const taskService = new TaskService();
